rasa/core/training/story_reader/yaml_story_reader.py

```diff
@@ -0,0 +1,245 @@
+import logging
+from typing import Dict, Text, List, Any
+
+import rasa.utils.io
+from rasa.constants import DOCS_URL_STORIES
+from rasa.core.constants import INTENT_MESSAGE_PREFIX
+from rasa.core.events import UserUttered, SlotSet
+from rasa.core.training.story_reader.story_reader import StoryReader
+from rasa.core.training.structures import StoryStep
+from rasa.data import YAML_FILE_EXTENSIONS
+from rasa.utils.common import raise_warning
+
+logger = logging.getLogger(__name__)
+
+KEY_STORIES = "stories"
+KEY_PLAIN_STORY_NAME = "story"
+KEY_PLAIN_STORY_STEPS = "steps"
+KEY_PLAIN_STORY_USER_UTTERANCE = "user"
+KEY_SLOT_NAME = "slot"
+KEY_SLOT_VALUE = "value"
+KEY_ACTION = "action"
+KEY_PLAIN_STORY_ENTITIES = "entities"
+KEY_CHECKPOINT = "checkpoint"
+KEY_CHECKPOINT_SLOTS = "slots"
+KEY_METADATA = "metadata"
+KEY_OR = "or"
+
+
+class YAMLStoryReader(StoryReader):
+    """Class that reads the core training data in a YAML format
+
+    """
+
+    async def read_from_file(self, filename: Text) -> List[StoryStep]:
+        try:
+            yaml_content = rasa.utils.io.read_yaml_file(filename)
+
+            stories = yaml_content.get(KEY_STORIES)  # pytype: disable=attribute-error
+            if not stories:
+                return []
+
+            self._parse_stories(stories)
+            self._add_current_stories_to_result()
+            return self.story_steps
+
+        except ValueError:
+            logger.error(f"Failed to read {filename}, it will be skipped")
+
+        return []
+
+    def _parse_stories(self, stories: List[Dict[Text, Any]]) -> None:
+        for story_item in stories:
+            if not isinstance(story_item, dict):
```

Comment:
this contradicts the parameter type of `stories` ` List[Dict[Text, Any]]`

Reply:
It's a list of dictionaries, and we iterate over the item, why does it contradict?
I think this check makes sense in case the user write the Yaml wrong.

Also I think it makes sense to have a schema here, I'll implement that later.

Reply:
>  why does it contradict?

Because the parameter says that we expect a `List[Dict]` and then we check if the list items are actually `dicts`.

Will we add yaml schemas later?

Reply:
> Because the parameter says that we expect a List[Dict] and then we check if the list items are actually dicts.

It's not a runtime check, that can still fail due to Yaml structure.

Reply:
So you mean that we will remove the check `isinstance(story_item, dict)` later when add schemas?

Reply:
I think so, yes. From what I see the schema can do these things for us.
I've just discovered that too late 😞 
Don't want to block you, so I can bring the schema later.

Reply:
sounds good 👍